oci-image/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+    ImageManifestFileWrite(String),
+    /// An error occurred when trying to open the state file
+    OpenStateFile(String),
+    /// An error occurred when trying to parse the state file
+    ParseStateFile(String),
+    /// An error occurred when trying to serialize the state
+    SerializeState(String),
+    /// An error occurred when trying to write the state into the file
+    WriteStateFile(String),
+    /// An error occurred when trying to parse the image configuration
+    ParseImageConfiguration(String),
+    /// The pulled number of layers is invalid
+    InvalidPulledLayers(String),
+    /// The layer uncompressed digest is different than the uncompressed layer digest defined in the image configuration.
+    UncompressedLayerInvalid(String),
+    /// The image is not found on the disk
+    ImageNotFound(String),
+    /// An error occurred when trying to mount OverlayFS layers
+    OverlayFSMount(String),
+    /// An error occurred when trying to umount an existing OverlayFS
+    OverlayFSUmount(nix::Error),
+    /// An error occurred when trying to create an OverlayFS related directory
+    OverlayFSCreateDirectory(String),
+    /// An error occurred when trying to generate the OCI config for a bundle
+    GenerateOCIConfig(oci_spec::OciSpecError),
+    /// An error occured when trying to unpack a layer
+    UnpackLayer(String),
+}
+
+/// A common result type for our crate.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The `ImageManager` should be responsible of the OCI images management.
+///
+/// It provides method to pull and unpack OCI images from OCI registries. It maintains a consistent state
+/// in order to avoid pulling existing layers, to improve performance.
+pub struct ImageManager {
+    /// The directory where images and bundles will be stored on the host.
+    data_dir: PathBuf,
+    /// The image manager state
+    state: Arc<Mutex<State>>,
+    /// The snapshot to use
+    snapshot: OverlayFS,
+}
+
+impl ImageManager {
+    pub fn new(data_dir: &Path) -> Result<Self> {
+        // If data_dir not is not existing
+        // we must create it
+        if !data_dir.exists() {
+            create_dir_all(data_dir)
+                .map_err(|e| Error::ManagerDataDirectoryCreation(e.to_string()))?;
+        }
+
+        let mut state = State::try_from(&data_dir.join(STATE_FILE))?;
+        let overlay_index = *state.get_snapshot("overlay").unwrap_or(&0);
+        let snapshot = OverlayFS {
+            data_dir: data_dir.join("overlay"),
+            index: AtomicUsize::new(overlay_index),
+        };
+
+        state.add_snapshot(overlay_index);
+
+        Ok(Self {
+            data_dir: data_dir.to_path_buf(),
+            state: Arc::new(Mutex::new(state.clone())),
+            snapshot,
+        })
+    }
+
+    /// Unpack an OCI image.
+    ///
+    /// Returns an error if the image is not found.
+    pub async fn mount(&mut self, image_id: &str, destination: &Option<PathBuf>) -> Result<String> {
```

Comment:
I think I don't understand what we want for this method/command. In my mind, If I wish to run a container with `kaps`, I will first pull the image : 

```
$ kaps pull docker.io/amd64/alpine --name alpine
``` 

Then, I have to create a bundle based on my image, so : 

```
$ kaps mount alpine /tmp/alpine
```
The path `/tmp/alpine` here is where we want to mount the container bundle. But if we simply use: `kaps mount alpine`, the bundle will be mounted in the kaps working directory (`/var/lib/kaps`).

Then, I can simply run my container with : 

```
$ kaps run --bundle /tmp/alpine # or the path returned by the mount if the user hasn't specified a specific path
```

Is this ok?

Reply:
It's pretty close to what I have in mind. My idea was simpler:

```
$ kaps pull docker.io/amd64/alpine # no --name
```

```
# Mount just returns the path to where kaps mounted it.
# If the caller needs it to be somewhere else, it can bind mount that path.
$ kaps mount docker.io/amd64/alpine
/var/lib/kaps/snapshot/<id>/
```

Reply:
Ok, I'll remove the `destination` parameter. 

The `--name` can be used to give an identifier for the image. Shall we remove it also?

Reply:
> The `--name` can be used to give an identifier for the image. Shall we remove it also?

As long as things work without the name, I'm fine. Note that adding support for `--name` adds code complexity that you could avoid for now.

Reply:
Without the name, the id of the image is computed by hashing the reference. By using that method, I'm sure that my identifier is unique. So yes, without name, it works perfectly also.